src/libs/fsdrv/lv_fs_littlefs.c

```diff
@@ -185,4 +185,76 @@ static lv_fs_res_t fs_tell(lv_fs_drv_t * drv, void * file_p, uint32_t * pos_p)
     return (int32_t)(*pos_p) < 0 ? LV_FS_RES_UNKNOWN : LV_FS_RES_OK;
 }
 
+/**
+ * Open a directory
+ * @param drv       pointer to a driver where this function belongs
+ * @param path      path to the directory beginning with the driver letter (e.g. S:/folder)
+ * @return          a directory descriptor or NULL on error
+ */
+static void * fs_dir_open(lv_fs_drv_t * drv, const char * path)
+{
+    LittleDirectory * ld = (LittleDirectory *)lv_malloc(sizeof(LittleDirectory));
+    LV_ASSERT_NULL(ld);
+
+    lfs_t * lfs = (lfs_t *)drv->user_data;
+    int err = lfs_dir_open(lfs, &ld->dir, path);
+    if(err != LFS_ERR_OK) {
+        return NULL;
+    }
+
+    return (void *)ld;
+}
+
+/**
+ * Close an opened directory
+ * @param drv      pointer to a driver where this function belongs
+ * @param dir_p    pointer to a dir_p variable. (opened with fs_dir_open)
+ * @return         LV_FS_RES_OK: no error or any error from @lv_fs_res_t enum
+ */
+static lv_fs_res_t fs_dir_close(lv_fs_drv_t * drv, void * dir_p)
+{
+    LittleDirectory * ld = (LittleDirectory *)dir_p;
+
+    lfs_t * lfs = (lfs_t *)drv->user_data;
+    int rc = lfs_dir_close(lfs, &ld->dir);
+
+    return rc < 0 ? LV_FS_RES_UNKNOWN : LV_FS_RES_OK;
+}
+
+/**
+ * Read data from an opened directory
+ * @param drv      pointer to a driver where this function belongs
+ * @param dir_p    pointer to a dir_p variable. (opened with fs_dir_open)
+ * @param fn       pointer to a buffer to store the filename
+ * @param fn_len   length of the buffer to store the filename
+ * @return         LV_FS_RES_OK: no error or any error from @lv_fs_res_t enum
+ */
+static lv_fs_res_t fs_dir_read(lv_fs_drv_t * drv, void * dir_p, char * fn, uint32_t fn_len)
+{
+    LittleDirectory * lf = (LittleDirectory *)dir_p;
+
+    lfs_t * lfs = (lfs_t *)drv->user_data;
+
+    int res;
+    struct lfs_info info;
+
+    fn[0] = '\0';
+    do {
+        res = lfs_dir_read(lfs, &lf->dir, &info);
+
+        if(info.name[0] == 0) break; /* End of the directory */
+
+        lv_strncpy(fn, info.name, fn_len);
```

Comment:
`lfs_dir_read` indicates the end of the dir by returning `0`.

From `lfs.h`:
```c
// Read an entry in the directory
//
// Fills out the info structure, based on the specified file or directory.
// Returns a positive value on success, 0 at the end of directory,
// or a negative error code on failure.
int lfs_dir_read(lfs_t *lfs, lfs_dir_t *dir, struct lfs_info *info);
```

Also, it's a convention for the lv fs drivers to add a `/` to the beginning if it's a directory.

For the strncpy change, see my comment in XuNeo's review.

```suggestion
        if(res < 0) return LV_FS_RES_UNKNOWN;
        if(res == 0) { /* End of the directory */
            fn[0] = '\0';
            break;
        }

        if(info.type != LFS_TYPE_DIR) {
            lv_strncpy(fn, info.name, fn_len - 1);
            fn[fn_len - 1] = '\0';
        }
        else {
            lv_snprintf(fn, fn_len, "/%s", info.name);
        }
```

Reply:
Woot, thanks for your awesome explanation, I better understand ^^
I'm gonna update this tomorrow morning :pray: